doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+blocks of code to the optimizer which provides more opportunities for other optimizations to improve the program.
+Driven by the expectation that the benefits of other optimizations such as constant folding, branch simplification,
+etc. that could occur, inlining enables JIT to do other powerful optimizations downstream of inlining.
+
+For example, if an inlined method passes a constant from the caller to the callee and the constant is used on
+the callee side, the code can be simplified dramatically by folding away the branches or changing loads into
+constants. Such as when passing a constant `String` from a caller to a callee and `hashCode()` or `equal()` is
+called on the constant `String` in the callee, these operations can be completely folded away during the compile
+time rather than walking the entire `String` and compute the hash code in runtime. For this `String`, the answer
+is already known during the compile time. Therefore, the call to `hashCade()` is replaced with a constant `String`.
+It is powerful because a call with a non-trivial logic is replaced with a simple constant.
+
+Similarly, if the callee compares the `String` that is passed in with some other constant strings, the comparison
+can also be performed during the compile time. The call `equal()` is completely eliminated in the runtime.
+
+Inlining also allows to stack allocate an object better because of a broader scope of things as a result of inlining.
+A return object from the callee to the caller could also be eliminated if the method is inlined. If the caller uses
+the object and then stops using it, it is possible to allocate the object on the stack. When the callee is brought
+into the caller, there is no need to save and restore registers across the call. Register assignment can also be
+done in a better way.
+
+Inlining is an enabler for other optimizations that run later. That is partially the reason why inliner runs
+fairly early in an optimization strategy. Code should be exposed as much as possible and as early as possible
+to other optimizations so that other optimizations can take advantage of what inlining has been done and optimize
+the code better.
+
+# 2. The Downside Of Inlining
+
+The downside is that the method gets much bigger if many callees are inlined into a method. In general, for a
+language like Java that has so many calls and is object oriented, it is always beneficial to inline a method on
+the hot path, unless the callee is really large and it executes lots of instructions. A small method on a hot
+path tends to get inlined in a language like Java because of so much calling and object orientedness built into it.
+
+# 3. Balance Multiple Paths Of Optimizations
+
+There is a division of responsibilities among the paths of optimizations in different parts of the JIT. We do not
+want to pollute every optimization with every other optimization. For example, we do not want to do folding or
```

Comment:
"with every other optimization" -> "with non-trivial cleanup tasks that can be done in another optimization"